source/language/pulses.rst

```diff
@@ -103,3 +102,159 @@ circuit. Most of the time symbols in the ``defcal`` table will also have
+
+.. code-block:: c
+
+   OPENQASM 3;
+   defcalgrammar "openpulse";
+
+
+   cal {
+      // declare global channel
+      channel d0 = getchannel("drive", $0);
+      // declare global frame
+      frame d0f = newframe(5.0e9, 0.0);
+
+   }
+
+   defcal x $0 {
+      waveform xp = gaussian(1.0, 160t, 40dt);
+      // References channel and frame declared in top-level cal block
+      play(d0, xp, d0f);
+   }
+
+
+Restrictions on defcal bodies
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The contents of ``defcal`` bodies are subject to the restriction they must have a definite length
+known at compile time, regardless of the parameters passed in or the state of the system when
+called. This allows the compiler to properly resolve ``lengthof(...)`` calls and
+allows for optimizations. If there is to be control flow in the ``defcal``, each branch of the
+control flow must have definite and equivalent length resolvable at compile time. Similarly, loops
+must be have a resolvable definite length at compile time.
+
+For example,  consider the case of a ``reset`` gate. The ``defcal`` for a
+``reset`` gate can be composed of a single if statement, provided each branch
+of the if statement has definite and equivalent length.
+
+.. code-block:: c
+
+   defcal reset $0 {
+      bit res = // measure qubit $0
+      if (res == 1) {
+         // flip the qubit
+      } else {
+         // delay for an equivalent amount of time
+      }
+   }
+
+Calibrations in practice
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+By their very nature calibrations are transient and unique to a target system.
+They are typically generated by automatic calibration routines that are periodically
+run on the target system, that are in turn bootstrapped from previous calibrations.
+The majority of OpenQASM users will use the default calibrations, however,
+for those that want more control, but do not want to bootstrap calibrations for an entire
+system it is expected that the target system provider will provide a ``backend.inc`` include
+file to the user. This will contain the declaration of the ``defcalgrammar``, constants,
+``defcal``s and other grammar and system specific components such as ``channel``s,
+``waveform``s and ``frame``s in the `OpenPulse defcalgrammar <openpulse.html>`. The user
+may than plugin to the existing calibrations by defining new calibrations, or overwriting
+existing ones by using the same ``channel``s and ``frame``s.
+The example below demonstrates this in practice for a two-qubit,
+cross-resonance device.
+
+.. code-block:: c
+
+   // backend.inc for openpulse two-qubit device
+
+   defcalgrammar "openpulse";
+
+   const q0_freq = 5.0e9;
+   const q1_freq = 5.1e9;
+
+   cal {
+
+      extern drag(complex[size] amp, length l, length sigma, float[size] beta) -> waveform;
+      extern gaussian_square(complex[size] amp, length l, length square_width, length sigma) -> waveform;
```

Comment:
By "we" do you mean the OpenQASM compiler?  Can you expound on the point?

Reply:
IMO, this falls into the same problem as the `capture` `extern`. So what if we require supporting a `durationof` call on an `extern`? For immediate functions it will trivially return 0, but for operations like `capture` or custom waveform templates, we may actually return a meaningful duration we need for real-time timekeeping/synchronization.

Reply:
This doesn't seem any different than the compiler having to handle the `capture` `extern`. Is it enough to say that `extern` symbols should support a `durationof` call? Most of these `extern`s will be trivially 0, but some like pulse templates or captures would return a meaningful result for scheduling.